### PR TITLE
community/SQLDatabase: Fetch mode `cursor`, query parameters, query by selectable, expose execution options, and documentation

### PR DESCRIPTION
**Pull request by amotl**
- **Description:** Improve `SQLDatabase` adapter component to promote code re-use, see [suggestion](https://github.com/langchain-ai/langchain/pull/16246#pullrequestreview-1846590962).
  - **Needed by:** GH-16246
  - **Addressed to:** @baskaryan, @cbornet 

## Details
- Add `cursor` fetch mode
- Accept SQL query parameters
- Accept both `str` and SQLAlchemy selectables as query expression
- Expose `execution_options`
- Documentation page (notebook) about `SQLDatabase` [^1]
  See [About SQLDatabase](https://github.com/langchain-ai/langchain/blob/c1c7b763/docs/docs/integrations/tools/sql_database.ipynb).

[^1]: Apparently there hasn't been any yet?

### CODE REVIEW

File: libs/langchain/tests/unit_tests/test_sql_database.py [hunk 1]
```diff
@@ -131,6 +132,51 @@ def test_sql_database_run() -> None:
...
+    with engine.begin() as conn:
+        conn.execute(stmt)
+    db = SQLDatabase(engine)
+    command = select(user).where(user.c.user_id == 17)
+
+    full_output = db.run(command, include_columns=True)
+    expected_full_output = "[{'user_id': 17, 'user_name': 'hwchase', 'user_bio': None}]"
+    assert full_output == expected_full_output
+
+
```
Comment (amotl) [inline libs/langchain/tests/unit_tests/test_sql_database.py:179]: I may modularize/generalize/compress this on a later iteration?

File: libs/community/langchain_community/utilities/sql_database.py [hunk 2]
```diff
@@ -373,14 +382,16 @@ def _get_sample_rows(self, table: Table) -> str:
 
     def _execute(
         self,
-        command: str,
-        fetch: Literal["all", "one"] = "all",
-    ) -> Sequence[Dict[str, Any]]:
+        command: Union[str, Executable],
+        fetch: Literal["all", "one", "result"] = "all",
```
Comment (amotl) [inline libs/community/langchain_community/utilities/sql_database.py:386]: Is there a spot in the documentation which would need to be updated along the lines, to educate users about the new `result` fetch mode?

File: libs/community/langchain_community/utilities/sql_database.py [hunk 3]
```diff
@@ -373,14 +382,16 @@ def _get_sample_rows(self, table: Table) -> str:
...
+        parameters: Optional[Dict[str, Any]] = None,
```
Comment (amotl) [inline libs/community/langchain_community/utilities/sql_database.py:387]: Dito. Should this new argument be documented appropriately?

File: libs/community/langchain_community/utilities/sql_database.py [hunk 4]
```diff
@@ -410,30 +421,45 @@ def _execute(
                 elif self.dialect == "postgresql":  # postgresql
                     connection.exec_driver_sql("SET search_path TO %s", (self._schema,))
 
-            cursor = connection.execute(text(command))
+            if isinstance(command, str):
+                command = text(command)
+            elif isinstance(command, Executable):
+                pass
+            else:
+                raise TypeError(f"Query expression has unknown type: {type(command)}")
```
Comment (amotl) [inline libs/community/langchain_community/utilities/sql_database.py:429]: That's the third feature of this patch. Same here: Should this capability be added to the documentation?

File: libs/langchain/tests/unit_tests/test_sql_database.py (hunk 1 above)
Comment (amotl) [reply libs/langchain/tests/unit_tests/test_sql_database.py]: GH-16659 will improve the situation.

File: libs/community/langchain_community/utilities/sql_database.py (hunk 2 above)
Comment (amotl) [reply libs/community/langchain_community/utilities/sql_database.py:386]: We have not been able to discover a dedicated documentation section for the `SQLDatabase` adapter utility.

> Up until now, the `SQLDatabase` utility only has been demonstrated on
> behalf of the "toolkit integration" section, where it is about
> showcasing an agent designed to interact with an SQL database. [1]
>
> Contrary to that, cf57a092483b now adds a documentation page exclusively
> dedicated to the `SQLDatabase` adapter utility, and its features.
> 
> [1] https://python.langchain.com/docs/integrations/toolkits/sql_database

We hope you like it. Please provide guidance if the content is inappropriate, needs to be changed, or slotted into a different directory. Thanks.

File: libs/community/langchain_community/utilities/sql_database.py (hunk 3 above)
Comment (amotl) [reply libs/community/langchain_community/utilities/sql_database.py:387]: Dito. Fixed with cf57a092483b.

File: libs/community/langchain_community/utilities/sql_database.py (hunk 4 above)
Comment (amotl) [reply libs/community/langchain_community/utilities/sql_database.py:429]: Dito. Fixed with cf57a092483b.

File: libs/community/langchain_community/utilities/sql_database.py [hunk 5]
```diff
@@ -410,30 +421,45 @@ def _execute(
...
+            cursor = connection.execute(command, parameters)
```
Comment (cbornet) [inline libs/community/langchain_community/utilities/sql_database.py:430]: Maybe also expose `execution_options` ?

Comment (amotl) [reply libs/community/langchain_community/utilities/sql_database.py]: Exposed per 696a52fc665.

File: libs/core/langchain_core/callbacks/manager.py
Comment (baskaryan) [inline libs/core/langchain_core/callbacks/manager.py:1]: why is this needed? feels like the on_text method should really just accept strings

File: libs/experimental/langchain_experimental/sql/vector_sql.py [hunk 6]
```diff
@@ -78,6 +78,7 @@ def parse(self, text: str) -> str:
 
 def get_result_from_sqldb(db: SQLDatabase, cmd: str) -> Sequence[Dict[str, Any]]:
     result = db._execute(cmd, fetch="all")
+    assert isinstance(result, Sequence)
```
Comment (baskaryan) [inline libs/experimental/langchain_experimental/sql/vector_sql.py:81]: why's this needed?

File: libs/community/langchain_community/utilities/sql_database.py [hunk 7]
```diff
@@ -373,67 +382,111 @@ def _get_sample_rows(self, table: Table) -> str:
 
     def _execute(
         self,
-        command: str,
-        fetch: Literal["all", "one"] = "all",
-    ) -> Sequence[Dict[str, Any]]:
+        command: Union[str, Executable],
+        fetch: Literal["all", "one", "result"] = "all",
```
Comment (baskaryan) [inline libs/community/langchain_community/utilities/sql_database.py:386]: "result" is a bit ambigious, would "cursor" or "cursor_result" be clearer?

File: libs/experimental/langchain_experimental/sql/vector_sql.py (hunk 6 above)
Comment (amotl) [reply libs/experimental/langchain_experimental/sql/vector_sql.py:81]: Hi @baskaryan. This adjustment was needed to satisfy the mypy linter, which would otherwise fail.

File: libs/community/langchain_community/utilities/sql_database.py (hunk 7 above)
Comment (amotl) [reply libs/community/langchain_community/utilities/sql_database.py:386]: I am fine with both of your suggestions. Let's go with just `cursor`?

File: libs/core/langchain_core/callbacks/manager.py
Comment (amotl) [reply libs/core/langchain_core/callbacks/manager.py:1]: Hi @baskaryan. This adjustment was needed to satisfy the mypy linter, which would otherwise fail.

File: libs/experimental/langchain_experimental/sql/vector_sql.py (hunk 6 above)
Comment (baskaryan) [reply libs/experimental/langchain_experimental/sql/vector_sql.py:81]: ah can we just cast the output of SQLDatabase wherever it's passed to on_text instead of changing the on_text method (since it doesn't actually support non-strings)? happy to do this if you want to give me write permissions to your fork

Comment (amotl) [reply libs/experimental/langchain_experimental/sql/vector_sql.py:81]: Sure we can do it, and I will also be happy to give you permissions so you can adjust the patch accordingly. I've just sent out an invite. Thank you very much!

File: libs/community/langchain_community/utilities/sql_database.py (hunk 7 above)
Comment (amotl) [reply libs/community/langchain_community/utilities/sql_database.py:386]: Will you also do the renaming now that you gained write permissions to this patch? I don't want to step on your toes from now on, unless you tell me.

Comment (amotl) [reply libs/community/langchain_community/utilities/sql_database.py:386]: I will do the renaming, to save you a few keystrokes of editing work.

Comment (amotl) [reply libs/community/langchain_community/utilities/sql_database.py]: Just pushed 9abeb24565fcc, which implements this suggestion. Thanks!

File: libs/core/langchain_core/callbacks/manager.py [hunk 8]
```diff
@@ -458,7 +458,7 @@ class RunManager(BaseRunManager):
 
     def on_text(
         self,
-        text: str,
+        text: Union[str, Any],
```
Comment (amotl) [inline libs/core/langchain_core/callbacks/manager.py]: > Ah can we just cast the output of SQLDatabase wherever it's passed to on_text instead of changing the on_text method (since it doesn't actually support non-strings)?

Undid this adjustment with bd7a515963.

File: libs/experimental/langchain_experimental/sql/vector_sql.py (hunk 6 above)
Comment (amotl) [reply libs/experimental/langchain_experimental/sql/vector_sql.py:81]: Hi again. Without that assert statement, `mypy` would trip like this:
```
$ cd libs/experimental
$ mypy langchain_experimental/sql/vector_sql.py
langchain_experimental/sql/vector_sql.py:82: error: Incompatible return value type (got "Union[Sequence[Dict[str, Any]], Result[Any]]", expected "Sequence[Dict[str, Any]]")  [return-value]
```
Do you have any idea how to resolve it differently?

NB: Regarding the type hinting change to the `on_text` method, see https://github.com/langchain-ai/langchain/pull/16655#pullrequestreview-1863481760.

File: libs/core/langchain_core/callbacks/manager.py
Comment (amotl) [reply libs/core/langchain_core/callbacks/manager.py:1]: > Can we just cast the output of SQLDatabase wherever it's passed to `on_text` instead of changing the `on_text` method (since it doesn't actually support non-strings)?

Hi again. I just reverted this adjustment, and added a cast at a different spot like you suggested, on behalf of 08db71bc. Let's see what CI will report about it.

Comment (amotl) [reply libs/core/langchain_core/callbacks/manager.py:1]: After rebasing this patch on top of `master`, it needed a few adjustments, which have been implemented now. Based on the CI outcome, the patch may receive another round of code reviews. Thank you.

File: libs/experimental/langchain_experimental/sql/base.py [hunk 9]
```diff
@@ -173,7 +173,7 @@ def _call(
                 sql_cmd = checked_sql_command
 
             _run_manager.on_text("\nSQLResult: ", verbose=self.verbose)
-            _run_manager.on_text(result, color="yellow", verbose=self.verbose)
+            _run_manager.on_text(str(result), color="yellow", verbose=self.verbose)
```
Comment (amotl) [inline libs/experimental/langchain_experimental/sql/base.py:176]: > Can we just cast the output of SQLDatabase wherever it's passed to `on_text` instead of changing the `on_text` method.

@baskaryan: Your suggestion has been implemented here.

File: libs/langchain/tests/unit_tests/test_sql_database.py
Comment (amotl) [inline libs/langchain/tests/unit_tests/test_sql_database.py:1]: I am wondering if this unit test file, together with [test_sql_database_schema.py](https://github.com/langchain-ai/langchain/blob/e4d2f3ea0bc6bd71aef71b3eef8263120567047f/libs/langchain/tests/unit_tests/test_sql_database_schema.py), should actually move into [community/tests/unit_tests](https://github.com/langchain-ai/langchain/tree/master/libs/community/tests/unit_tests)?

**Comment by eyurtsev** [comment]
Reviewing and comandeering to adjust a few things in the API

File: libs/experimental/langchain_experimental/sql/base.py (hunk 9 above)
Comment (eyurtsev) [reply libs/experimental/langchain_experimental/sql/base.py:176]: This seems fine to me

File: libs/langchain/tests/unit_tests/test_sql_database.py
Comment (eyurtsev) [reply libs/langchain/tests/unit_tests/test_sql_database.py:1]: Probably yes, we can follow up in a separate PR

**Comment by eyurtsev** [comment]
@amotl are you able to give me permissions to modify the head on the given branch? (it's an option you can enable when creating the PR)

I fixed a breaking change due to lack of `*` syntax, and updated the notebook with a security notebook. We can merge afterwards.

I left another review on the original PR to see if keyset pagination is an option (instead of relying on a cursor)

**Comment by eyurtsev** [comment]
Here is a PR that contains the proposed changes: https://github.com/langchain-ai/langchain/pull/17191/files, would prefer to push here so you can :+1:  :-1:  the differences

File: libs/experimental/langchain_experimental/sql/vector_sql.py (hunk 6 above)
Comment (amotl) [reply libs/experimental/langchain_experimental/sql/vector_sql.py:81]: Hi @eyurtsev. Thanks for the review.

Down this conversation, you asked:

> @amotl are you able to give me permissions to modify the head on the given branch? (it's an option you can enable when creating the PR)

Is there a way to modify it on the PR afterwards? I admit I never recognized that option, apologies. In any case, I've added you to the list of contributors, where you may be able to accept it on the [invitations page](https://github.com/crate-workbench/langchain/invitations). Does that help?

**Comment by amotl** [comment]
Hi @eyurtsev. Thanks for the review.

> are you able to give me permissions to modify the head on the given branch? (it's an option you can enable when creating the PR)

Is there a way to modify it on the PR afterwards? I admit I never recognized that option, apologies. In any case, I've added you to the list of contributors, where you may be able to accept it on the [invitations page](https://github.com/crate-workbench/langchain/invitations). Does that help already?

> Here is a PR that contains the proposed changes: https://github.com/langchain-ai/langchain/pull/17191/files, would prefer to push here so you can 👍 👎 the differences

Thanks. I've looked at 4f3b0630666a9aded3ed10fe1bdca27fafe2c831 and 806ba2ee4d483d94d42c3fb4b4af8422a9edff33, and I appreciate your improvements. Now, I hope we will get the permission problem sorted out?

File: libs/core/langchain_core/callbacks/manager.py
Comment (amotl) [reply libs/core/langchain_core/callbacks/manager.py:1]: Hi again. It looks like @eyurtsev is taking over. Thank you so much!

**Comment by amotl** [comment]
> I hope we will get the permission problem sorted out?

I've just digested the documentation about [Allowing changes to a pull request branch created from a fork](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork), but I don't see the **Allow edits from maintainers.** or **Allow edits and access to secrets by maintainers.** options on the PR.

Within the documentation, they refer to "forks owned by a personal account". Is that the reason it doesn't work, because this fork is on an organization account?

**Comment by eyurtsev** [comment]
@amotl I'll close this PR and merge the branch I had: https://github.com/langchain-ai/langchain/pull/17191 hope this is OK

**Comment by eyurtsev** [comment]
Merged the other PR in. I'll read up on the links you sent to try and figure out if there's any other way to do this -- from what i understood it sounds like there's a difference between forks owned by individuals vs. forks owned by orgs

**Comment by amotl** [comment]
> Merged the other PR in. Hope this is OK.

Definitively, I appreciate it. Thanks!

> from what i understood it sounds like there's a difference between forks owned by individuals vs. forks owned by orgs

Exactly. Apparently, repos owned by orgs do not offer that feature to permit upstream authors access to their PRs.
